startup/gui/performanceMonitor.py

```diff
@@ -0,0 +1,127 @@
+#        the distribution.
+#
+#      * Neither the name of Cinesite VFX Ltd. nor the names of
+#        any other contributors to this software may be used to endorse or
+#        promote products derived from this software without specific prior
+#        written permission.
+#
+#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+#  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+#  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+#  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+#  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+#  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+#  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+#  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+#  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+#  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+#  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+##########################################################################
+
+import IECore
+
+import Gaffer
+import GafferUI
+
+menu = GafferUI.ScriptWindow.menuDefinition( application )
+
+__performanceMonitor = None
+
+def __startPerformanceMonitor() :
+
+	global __performanceMonitor
+	assert( __performanceMonitor is None )
+	__performanceMonitor = Gaffer.PerformanceMonitor()
+	__performanceMonitor.__enter__()
+
+def __stopPerformanceMonitor( menu ) :
+
+	global __performanceMonitor
+	__performanceMonitor.__exit__( None, None, None )
+
+	script = menu.ancestor( GafferUI.ScriptWindow ).scriptNode()
+
+	Gaffer.MonitorAlgo.annotate( script, __performanceMonitor, Gaffer.MonitorAlgo.PerformanceMetric.TotalDuration, persistent = False )
+	Gaffer.MonitorAlgo.annotate( script, __performanceMonitor, Gaffer.MonitorAlgo.PerformanceMetric.HashCount, persistent = False )
+	Gaffer.MonitorAlgo.annotate( script, __performanceMonitor, Gaffer.MonitorAlgo.PerformanceMetric.ComputeCount, persistent = False )
+
+	__performanceMonitor = None
+
+__contextMonitor = None
+
+def __startContextMonitor( menu ) :
+
+	global __contextMonitor
+	assert( __contextMonitor is None )
+
+	script = menu.ancestor( GafferUI.ScriptWindow ).scriptNode()
+
+	__contextMonitor = Gaffer.ContextMonitor( script.selection()[-1] )
+	__contextMonitor.__enter__()
+
+def __stopContextMonitor( menu ) :
+
+	global __contextMonitor
+	__contextMonitor.__exit__( None, None, None )
+
+	script = menu.ancestor( GafferUI.ScriptWindow ).scriptNode()
+	Gaffer.MonitorAlgo.annotate( script, __contextMonitor, persistent = False )
+
+	__contextMonitor = None
+
+def __profilingSubMenu( menu ) :
+
+	global __performanceMonitor
+
+	result = IECore.MenuDefinition()
+
+	result.append(
+		"/Performance Monitor/Start",
+		{
+			"command" : __startPerformanceMonitor,
+			"active" : __performanceMonitor is None
+		}
+	)
```

Comment:
First impressions from some playing around with this:
* If I didn't already know how it worked, I'm not sure I would have guessed that you have to press "Stop" in order to get results
* How do I clear out existing results if I've run one test and want to run another?  Currently results are just summed up?
* It's a bit confusing that there are metrics like "Hashes Per Compute" which are listed in the Annotation visibility menu, but don't seem to be created by running a performance monitor from this menu.